src/main/scala/is/hail/utils/richUtils/RichRDD.scala

```diff
@@ -99,4 +97,40 @@ class RichRDD[T](val r: RDD[T]) extends AnyVal {
         r.compute(split.asInstanceOf[SubsetRDDPartition].parentPartition, context)
     }
   }
-}
+
+  def headPerPartition(n: Int)(implicit ct: ClassTag[T]): RDD[T] = {
+    require(n >= 0)
+    r.mapPartitions({ it => it.take(n) }, preservesPartitioning = true)
+  }
+
+  def head(n: Long)(implicit ct: ClassTag[T]): RDD[T] = {
+    require(n >= 0)
+
+    val sc = r.sparkContext
+    val parallelism = sc.defaultParallelism
+    val nPartitions = r.partitions.length
+
+    val counts = new Array[Long](nPartitions)
+    var nScanned = 0
+    var nSeen = 0L
+
+    while (nSeen < n && nScanned < nPartitions) {
+      val scanPartitions = nScanned until (nScanned + parallelism): Seq[Int]
+      sc.runJob(r, getIteratorSize _, scanPartitions)
+        .zipWithIndex.foreach { case (c, i) =>
+        counts(nScanned + i) = c
+        nSeen += c
+      }
+      nScanned += parallelism
+    }
+
+    var nRemaining = n
```

Comment:
This stuff seems too complicated for what we're doing.
Conceptually, it seems all we need to figure out are:
 - which partition we stop on
 - how many elements to take from that last partition

Then, all we need to do is leave other partitions alone and take a subset from the last.